ethkey/src/extended.rs

```diff
@@ -0,0 +1,368 @@
+		let extended_secret = ExtendedSecret::with_code(secret, chain_code);
+		let extended_public = ExtendedPublic::from_secret(&extended_secret)
+			.expect("Valid `Secret` always produces valid public; qed");
+		ExtendedKeyPair {
+			secret: extended_secret,
+			public: extended_public,
+		}
+	}
+
+	pub fn secret(&self) -> &ExtendedSecret {
+		&self.secret
+	}
+
+	pub fn public(&self) -> &ExtendedPublic {
+		&self.public
+	}
+
+	pub fn derive(&self, index: u32) -> Result<Self, DerivationError> {
+		let derived = self.secret.derive(index);
+
+		Ok(ExtendedKeyPair {
+			public: ExtendedPublic::from_secret(&derived)?,
+			secret: derived,
+		})
+	}
+}
+
+// Derivation functions for private and public keys
+// Work is based on BIP0032
+// https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+mod derivation {
+
+	use rcrypto::hmac::Hmac;
+	use rcrypto::mac::Mac;
+	use rcrypto::sha2::Sha512;
+	use bigint::hash::{H512, H256, FixedHash};
+	use bigint::prelude::{U256, U512, Uint};
+	use byteorder::{BigEndian, ByteOrder};
+	use secp256k1;
+	use secp256k1::key::{SecretKey, PublicKey};
+	use SECP256K1;
+	use keccak;
+
+	#[derive(Debug)]
+	pub enum Error {
+		InvalidHardenedUse,
+		InvalidPoint,
+		MissingIndex,
+	}
+
+	// Deterministic derivation of the key using secp256k1 elliptic curve.
+	// Derivation can be either hardened or not.
+	// For hardened derivation, pass index at least 2^31
+	pub fn private(private_key: H256, chain_code: H256, index: u32) -> (H256, H256) {
+		if index < (2 << 30) {
+			private_soft(private_key, chain_code, index)
+		}
+		else {
+			private_hard(private_key, chain_code, index)
+		}
+	}
+
+	fn hmac_pair(data: [u8; 37], private_key: H256, chain_code: H256) -> (H256, H256) {
+		let private: U256 = private_key.into();
+
+		// produces 512-bit derived hmac (I)
+		let mut hmac = Hmac::new(Sha512::new(), &*chain_code);
+		let mut i_512 = [0u8; 64];
+		hmac.input(&data[..]);
+		hmac.raw_result(&mut i_512);
+
+		// left most 256 bits are later added to original private key
+		let hmac_key: U256 = H256::from_slice(&i_512[0..32]).into();
+		// right most 256 bits are new chain code for later derivations
+		let next_chain_code = H256::from(&i_512[32..64]);
+
+		let child_key = private_add(hmac_key, private).into();
+		(child_key, next_chain_code)
+	}
+
+	fn private_soft(private_key: H256, chain_code: H256, index: u32) -> (H256, H256) {
+		let mut data = [0u8; 37];
+
+		let sec_private = SecretKey::from_slice(&SECP256K1, &*private_key)
+			.expect("Caller should provide valid private key");
+		let sec_public = PublicKey::from_secret_key(&SECP256K1, &sec_private)
+			.expect("Caller should provide valid private key");
+		let public_serialized = sec_public.serialize_vec(&SECP256K1, true);
+
+		// curve point (compressed public key) --  index
+		//             0.33                    --  33..37
+		data[0..33].copy_from_slice(&public_serialized);
+		BigEndian::write_u32(&mut data[33..37], index);
```

Comment:
yup, but that number is theoretically searchable. my understanding is that the maths allows for `2**256` soft keys, which then means that you can have cryptographically secure receive addresses. basically, you want to make a secret payment to address `X`: you generate address `Y = soft(X, H)` where `H` is some cryptographically secure invoice identifier (e.g. the `sha3` of the invoice PDF). if you don't have `H` then you cannot associate X with Y, but once you do have `H` then you can prove that the owner of `X` can access the funds at `Y`.

basically it lets you send funds to someone's public address without the payment itself being made public. but still with the ability for either party to make the payment public at a later date.